File: Example/SnapshotTests/BPKMapAnnotationSnapshotTest.m
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright © 2019 Skyscanner Ltd. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <Backpack/Spacing.h>
#import <Backpack/Map.h>

#import <FBSnapshotTestCase/FBSnapshotTestCase.h>
#import "BPKSnapshotTest.h"

NSString *const ReuseIdentifier = @"Annotation";

@interface BPKMapAnnotationSnapshotTest : FBSnapshotTestCase
@end

NS_ASSUME_NONNULL_BEGIN
@implementation BPKMapAnnotationSnapshotTest

- (void)setUp {
    [super setUp];
    self.recordMode = NO;
}

-(UIView *)createMapAnnotationWithLabel:(NSString *)label alwaysShowCallout:(BOOL)alwaysShowCallout previouslySelected:(BOOL)previouslySelected selected:(BOOL)selected enabled:(BOOL)enabled icon:(BOOL)icon {
    BPKMapAnnotation *annotation =  [[BPKMapAnnotation alloc] init];
    annotation.title = label;
    annotation.alwaysShowCallout = alwaysShowCallout;
    annotation.enabled = enabled;
    if (icon) {
        annotation.iconName = BPKIconNameAirports;
    }

    BPKMapAnnotationView *annotationView = [[BPKMapAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:ReuseIdentifier];
    annotationView.selected = previouslySelected;
    annotationView.selected = selected;

    UIView *wrapper = [[UIView alloc] initWithFrame:CGRectZero];
    [wrapper addSubview:annotationView];
    annotationView.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [annotationView.widthAnchor constraintEqualToConstant:annotationView.bounds.size.width],
        [annotationView.heightAnchor constraintEqualToConstant:annotationView.bounds.size.height],
        [annotationView.centerXAnchor constraintEqualToAnchor:wrapper.centerXAnchor],
        [annotationView.centerYAnchor constraintEqualToAnchor:wrapper.centerYAnchor],
        [wrapper.widthAnchor constraintEqualToAnchor:annotationView.widthAnchor constant:BPKSpacingSm * 2],
        [wrapper.heightAnchor constraintEqualToAnchor:annotationView.heightAnchor constant:BPKSpacingSm * 2]
    ]];
    return wrapper;
}

-(UIView *)createMapAnnotationsWithIcons:(BOOL)icon {
    UIView *annotationView1 = [self createMapAnnotationWithLabel:@"Edinburgh" alwaysShowCallout:NO previouslySelected:NO selected:NO enabled:YES icon:icon];
    UIView *annotationView2 = [self createMapAnnotationWithLabel:@"Edinburgh" alwaysShowCallout:NO previouslySelected:NO selected:YES enabled:YES icon:icon];
    UIView *annotationView3 = [self createMapAnnotationWithLabel:@"Edinburgh" alwaysShowCallout:YES previouslySelected:NO selected:NO enabled:YES icon:icon];
    UIView *annotationView4 = [self createMapAnnotationWithLabel:@"Edinburgh" alwaysShowCallout:YES previouslySelected:YES selected:NO enabled:YES icon:icon];
    UIView *annotationView5 = [self createMapAnnotationWithLabel:@"Edinburgh" alwaysShowCallout:YES previouslySelected:NO selected:YES enabled:YES icon:icon];
    UIView *annotationView6 = [self createMapAnnotationWithLabel:@"Edinburgh" alwaysShowCallout:YES previouslySelected:NO selected:NO enabled:NO icon:icon];

    UIView *outerView = [[UIView alloc] initWithFrame:CGRectZero];

    [outerView addSubview:annotationView1];
    [outerView addSubview:annotationView2];
    [outerView addSubview:annotationView3];
    [outerView addSubview:annotationView4];
    [outerView addSubview:annotationView5];
    [outerView addSubview:annotationView6];

    outerView.translatesAutoresizingMaskIntoConstraints = NO;
    annotationView1.translatesAutoresizingMaskIntoConstraints = NO;
    annotationView2.translatesAutoresizingMaskIntoConstraints = NO;
    annotationView3.translatesAutoresizingMaskIntoConstraints = NO;
    annotationView4.translatesAutoresizingMaskIntoConstraints = NO;
    annotationView5.translatesAutoresizingMaskIntoConstraints = NO;
    annotationView6.translatesAutoresizingMaskIntoConstraints = NO;

    [NSLayoutConstraint activateConstraints:@[
        [annotationView1.leadingAnchor constraintEqualToAnchor:outerView.leadingAnchor],
        [annotationView1.topAnchor constraintEqualToAnchor:outerView.topAnchor],
        [outerView.bottomAnchor constraintGreaterThanOrEqualToAnchor:annotationView1.bottomAnchor],

        [annotationView2.leadingAnchor constraintEqualToAnchor:annotationView1.trailingAnchor],
        [annotationView2.topAnchor constraintEqualToAnchor:outerView.topAnchor],
        [outerView.bottomAnchor constraintGreaterThanOrEqualToAnchor:annotationView2.bottomAnchor],

        [annotationView3.leadingAnchor constraintEqualToAnchor:annotationView2.trailingAnchor],
        [annotationView3.topAnchor constraintEqualToAnchor:outerView.topAnchor],
        [outerView.bottomAnchor constraintGreaterThanOrEqualToAnchor:annotationView3.bottomAnchor],

        [annotationView4.leadingAnchor constraintEqualToAnchor:annotationView3.trailingAnchor],
        [annotationView4.topAnchor constraintEqualToAnchor:outerView.topAnchor],
        [outerView.bottomAnchor constraintGreaterThanOrEqualToAnchor:annotationView4.bottomAnchor],

        [annotationView5.leadingAnchor constraintEqualToAnchor:annotationView4.trailingAnchor],
        [annotationView5.topAnchor constraintEqualToAnchor:outerView.topAnchor],
        [outerView.bottomAnchor constraintGreaterThanOrEqualToAnchor:annotationView5.bottomAnchor],

        [annotationView6.leadingAnchor constraintEqualToAnchor:annotationView5.trailingAnchor],
        [annotationView6.topAnchor constraintEqualToAnchor:outerView.topAnchor],
        [outerView.bottomAnchor constraintGreaterThanOrEqualToAnchor:annotationView6.bottomAnchor],

        [outerView.trailingAnchor constraintEqualToAnchor:annotationView6.trailingAnchor]
    ]];
    return outerView;
}

- (void)testMapAnnotations {
    UIView *lightView = [self createMapAnnotationsWithIcons:NO];
    UIView *darkView = [self createMapAnnotationsWithIcons:NO];

    BPKSnapshotVerifyViewLight(lightView, nil);
    BPKSnapshotVerifyViewDark(darkView, nil);
}

- (void)testMapAnnotationsWithIcons {
    UIView *lightView = [self createMapAnnotationsWithIcons:YES];
    UIView *darkView = [self createMapAnnotationsWithIcons:YES];

    BPKSnapshotVerifyViewLight(lightView, nil);
    BPKSnapshotVerifyViewDark(darkView, nil);
}

-(UIView *)createLongTitleCalloutViewWithIcon:(BOOL)icon {
    UIView *annotationView = [self
                              createMapAnnotationWithLabel:@"Callout with a super super super long title that goes on and on and on"
                              alwaysShowCallout:YES
                              previouslySelected:NO
                              selected:NO
                              enabled:YES
                              icon:icon
                              ];
    annotationView.translatesAutoresizingMaskIntoConstraints = NO;
    return annotationView;
}

- (void)testLongTitleCalloutView {
    UIView *lightView = [self createLongTitleCalloutViewWithIcon:NO];
    UIView *darkView = [self createLongTitleCalloutViewWithIcon:NO];

    BPKSnapshotVerifyViewLight(lightView, nil);
    BPKSnapshotVerifyViewDark(darkView, nil);
}

- (void)testLongTitleCalloutViewWithIcon {
    UIView *lightView = [self createLongTitleCalloutViewWithIcon:YES];
    UIView *darkView = [self createLongTitleCalloutViewWithIcon:YES];

    BPKSnapshotVerifyViewLight(lightView, nil);
    BPKSnapshotVerifyViewDark(darkView, nil);
}

@end
NS_ASSUME_NONNULL_END

File: Backpack/Tests/SnapshotTests/MapAnnotation/BPKMapAnnotationSnapshotTest.swift
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018-2022 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import XCTest
import Backpack

import SnapshotTesting

class BPKMapAnnotationSnapshotTest: XCTestCase {
    
    private let reuseIdentifier = "Annotation"
    private let traitDarkMode = UITraitCollection(userInterfaceStyle: .dark)

    override func setUpWithError() throws {
        try? super.setUpWithError()
        isRecording = false
    }
    
    // MARK: State testing
    private func createAllStates(hasIcon: Bool) -> UIView {
        let stackView = createStackView()
        let parentView = UIView(frame: .zero)
        
        parentView.translatesAutoresizingMaskIntoConstraints = false
        parentView.addSubview(stackView)
        
        let label = "Edinburgh"
        
        let allStateViews = [
            create(withLabel: label, alwaysShowCallout: false, previouslySelected: false, selected: false, enabled: true, icon: hasIcon),
            create(withLabel: label, alwaysShowCallout: false, previouslySelected: false, selected: true, enabled: true, icon: hasIcon),
            create(withLabel: label, alwaysShowCallout: true, previouslySelected: false, selected: false, enabled: true, icon: hasIcon),
            create(withLabel: label, alwaysShowCallout: true, previouslySelected: true, selected: false, enabled: true, icon: hasIcon),
            create(withLabel: label, alwaysShowCallout: true, previouslySelected: false, selected: true, enabled: true, icon: hasIcon),
            create(withLabel: label, alwaysShowCallout: true, previouslySelected: false, selected: false, enabled: false, icon: hasIcon)
        ]
        
        allStateViews.forEach(stackView.addArrangedSubview(_:))
        
        NSLayoutConstraint.activate([
            stackView.topAnchor.constraint(equalTo: parentView.topAnchor),
            stackView.leadingAnchor.constraint(equalTo: parentView.leadingAnchor),
            parentView.bottomAnchor.constraint(equalTo: stackView.bottomAnchor),
            parentView.trailingAnchor.constraint(equalTo: stackView.trailingAnchor)
        ])
        
        return parentView
    }
    
    func testAllStates() {
        let exampleView = createAllStates(hasIcon: false)
        
        assertSnapshot(matching: exampleView, as: .image)
        assertSnapshot(matching: exampleView, as: .image(traits: traitDarkMode))
    }
    
    func testAllStatesWithIcon() {
        let exampleView = createAllStates(hasIcon: true)
        
        assertSnapshot(matching: exampleView, as: .image)
        assertSnapshot(matching: exampleView, as: .image(traits: traitDarkMode))
    }
    
    
    // MARK: Long titles
    private func createWithLongTitle(hasIcon: Bool) -> UIView {
        let view = create(
            withLabel: "Callout with a super super super long title that goes on and on and on",
            alwaysShowCallout: true,
            previouslySelected: false,
            selected: false,
            enabled: true,
            icon: hasIcon
        )
        
        return view
    }
    
    func testLongTitleCalloutViewWith() {
        let exampleView = createWithLongTitle(hasIcon: false)
        
        assertSnapshot(matching: exampleView, as: .image)
        assertSnapshot(matching: exampleView, as: .image(traits: traitDarkMode))
    }

    func testLongTitleCalloutViewWithIcon() {
        let exampleView = createWithLongTitle(hasIcon: true)
        
        assertSnapshot(matching: exampleView, as: .image)
        assertSnapshot(matching: exampleView, as: .image(traits: traitDarkMode))
    }
    
    // MARK: Util
    // swiftlint:disable:next function_parameter_count
    private func create(
        withLabel label: String,
        alwaysShowCallout: Bool,
        previouslySelected: Bool,
        selected: Bool,
        enabled: Bool,
        icon: Bool
    ) -> UIView {
        let annotation = BPKMapAnnotation()
        annotation.title = label
        annotation.alwaysShowCallout = alwaysShowCallout
        annotation.enabled = enabled
        
        if icon {
            annotation.iconName = .airports
        }
        
        let annotationView = BPKMapAnnotationView(
            annotation: annotation,
            reuseIdentifier: reuseIdentifier
        )
        
        annotationView.isSelected = previouslySelected
        annotationView.isSelected = selected
        
        let wrapper = UIView()
        wrapper.addSubview(annotationView)
        
        annotationView.translatesAutoresizingMaskIntoConstraints = false
        wrapper.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            annotationView.widthAnchor.constraint(equalToConstant: annotationView.bounds.size.width),
            annotationView.heightAnchor.constraint(equalToConstant: annotationView.bounds.size.height),
            annotationView.centerXAnchor.constraint(equalTo: wrapper.centerXAnchor),
            annotationView.centerYAnchor.constraint(equalTo: wrapper.centerYAnchor),
            wrapper.widthAnchor.constraint(equalTo: annotationView.widthAnchor, constant: BPKSpacingSm * 2),
            wrapper.heightAnchor.constraint(equalTo: annotationView.heightAnchor, constant: BPKSpacingSm * 2)
        ])
        
        return wrapper
    }
    
    private func createStackView() -> UIStackView {
        let stackView = UIStackView(frame: .zero)
        stackView.axis = .horizontal
        stackView.alignment = .leading
        stackView.distribution = .equalCentering
        stackView.spacing = BPKSpacingMd
        stackView.translatesAutoresizingMaskIntoConstraints = false
        return stackView
    }
}
